cpp/FileTransferAgent.cpp

```diff
@@ -259,12 +272,16 @@ void Snowflake::Client::FileTransferAgent::upload(string *command)
 
   if (m_smallFilesMeta.size() > 0)
   {
-    if (m_storageClient->requirePresignedUrl())
+    //skip getting presigned url if fast fail is enabled and failed already
+    if ((!isPutFastFailEnabled()) || m_failedTransfers.empty())
     {
-      for (size_t i = 0; i < m_smallFilesMeta.size(); i++)
+      if (m_storageClient->requirePresignedUrl())
       {
-        std::string localCommand = *command;
-        getPresignedUrlForUploading(m_smallFilesMeta[i], localCommand);
+        for (size_t i = 0; i < m_smallFilesMeta.size(); i++)
+        {
+          std::string localCommand = *command;
+          getPresignedUrlForUploading(m_smallFilesMeta[i], localCommand);
+        }
       }
     }
     uploadFilesInParallel(command);
```

Comment:
If fast fail is enabled and we've had a failure it would mean that we won't call `getPresignedUrlForUploading` from the logic between line 278 and 284. If that is the case should we even call `uploadFilesInParallel` at all?

Reply:
In such case, calling uploadFilesInParallel() is just for setting transfer outcome to SKIP_UPLOAD_FILE for following files. We could add a loop here to do that but I think it doesn't hurt to reuse the existing logic.